src/telemetry/stats.c

```diff
@@ -0,0 +1,579 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <access/table.h>
+#include <access/genam.h>
+#include <access/tableam.h>
+#include <access/htup_details.h>
+#include <catalog/pg_class.h>
+#include <catalog/indexing.h>
+#include <catalog/namespace.h>
+#include <catalog/pg_namespace.h>
+#include <utils/builtins.h>
+#include <fmgr.h>
+
+#include "stats.h"
+#include "ts_catalog/catalog.h"
+#include "ts_catalog/continuous_agg.h"
+#include "chunk.h"
+#include "extension.h"
+#include "hypertable_cache.h"
+#include "utils.h"
+
+typedef struct StatsContext
+{
+	TelemetryStats *stats;
+	ScanIterator compressed_chunk_stats_iterator;
+	bool iterator_valid;
+} StatsContext;
+
+/*
+ * Determine the type of a hypertable.
+ */
+static StatsRelType
+classify_hypertable(const Hypertable *ht)
+{
+	if (TS_HYPERTABLE_IS_INTERNAL_COMPRESSION_TABLE(ht))
+	{
+		/* This is an internal compression table, but could be for a
+		 * regular hypertable, a distributed member hypertable, or for
+		 * an internal materialized hypertable (cagg). The latter case
+		 * is currently not handled */
+		return RELTYPE_COMPRESSION_HYPERTABLE;
+	}
+	else
+	{
+		/* Not dealing with an internal compression hypertable, but
+		 * could be a materialized hypertable (cagg) unless it is
+		 * distributed.  */
+		switch (ht->fd.replication_factor)
+		{
+			case HYPERTABLE_DISTRIBUTED_MEMBER:
+				return RELTYPE_DISTRIBUTED_HYPERTABLE_MEMBER;
+			case HYPERTABLE_REGULAR:
+			{
+				const ContinuousAgg *cagg = ts_continuous_agg_find_by_mat_hypertable_id(ht->fd.id);
+
+				if (cagg)
+					return RELTYPE_MATERIALIZED_HYPERTABLE;
+
+				return RELTYPE_HYPERTABLE;
+			}
+			default:
+				Assert(ht->fd.replication_factor >= 1);
+				return RELTYPE_DISTRIBUTED_HYPERTABLE;
+		}
+	}
+}
+
+static StatsRelType
+classify_chunk(Cache *htcache, const Hypertable **ht, const Chunk *chunk)
```

Comment:
Good point. We do classify things in the planner (`src/planner.c`). That code is not as general as this one, but it might be worth looking into how to unify those code paths into something more general. I think we can leave that for follow-up work though.